Easily create custom form operations (using HasForm trait and general form view)

tabacitu opened the pull request:
## WHY

Note: This PR points to https://github.com/Laravel-Backpack/CRUD/pull/5118 - it needs that to be merged first.

### BEFORE - What was wrong? What was happening before this PR?

It was clunky and difficult to create custom operations that included Backpack forms. We have [a tutorial for it](https://backpackforlaravel.com/articles/tutorials/how-to-create-a-custom-operation-that-uses-backpack-fields) but it's a multi-step process, and after following it you still end up with non-working code that you need to fix yourself. 

### AFTER - What is happening after this PR?

It is MUUUUCH easier to create custom operations that include Backpack forms. Especially the most common ones:

**Step 1.** Create a new operation trait using `php artisan backpack:crud-operation Email`. 

> Note: This can be made even easier, by creating a new command with a stub specifically for this kind of operations: `php artisan backpack:crud-form-operation Email` (TODO - could/should).

**Step 2.** In the `CrudController` where you want that operation, use this new operation trait and add the fields you want in its setup method:

```php
class UserCrudController extends CrudController
{
    // ...
    use \App\Http\Controllers\Admin\Operations\EmailOperation;
    // ...

    protected function setupEmailOperation()
    {
        CRUD::field('to')->type('email')->size(4);
        CRUD::field('cc')->type('text')->size(4);
        CRUD::field('bcc')->type('text')->size(4);
        CRUD::field('subject')->type('text');
        CRUD::field('message')->type('textarea');
    }
```

**Step 3.** Inside `EmailOperation.php` use the brand-new `HasForm` trait and the methods it provides:

```php
<?php

namespace App\Http\Controllers\Admin\Operations;

use Illuminate\Support\Str;
use Backpack\CRUD\app\Library\CrudPanel\CrudPanelFacade as CRUD;
use Backpack\CRUD\app\Http\Controllers\Operations\Concerns\HasForm;

trait EmailOperation
{
    use HasForm;

    /**
     * Define which routes are needed for this operation.
     *
     * @param string $segment    Name of the current entity (singular). Used as first URL segment.
     * @param string $routeName  Prefix of the route name.
     * @param string $controller Name of the current CrudController.
     */
    protected function setupEmailRoutes($segment, $routeName, $controller)
    {
        $operationName = Str::of(__FUNCTION__)->after('setup')->before('Routes');
        $routesHaveIdSegment = true;

        $this->formRoutes($operationName, $routesHaveIdSegment, $segment, $routeName, $controller);
        // this will set up two routes:
        // - GET that points to getEmailForm()
        // - POST that points to postEmailForm()
    }

    /**
     * Add the default settings, buttons, etc that this operation needs.
     */
    protected function setupEmailDefaults()
    {
        // Access
        CRUD::allowAccess('Email');

        // Config
        $this->crud->operation($this->crud->getCurrentOperation(), function () {
            $this->crud->loadDefaultOperationSettingsFromConfig();
            $this->crud->setupDefaultSaveActions();
        });

        // Button
        $this->crud->operation('list', function () {
            $this->crud->button('email')->stack('line')->view('crud::buttons.quick');

            // this will add a quick line button, but you can also
            // configure how it looks and works by passing parameters:
            // $this->crud->button('email')->stack('line')->view('crud::buttons.quick')->meta([
                // 'access' => 'Email',
                // 'label' => 'Email',
                // 'icon' => 'la la-envelope',
                // 'classes' => '',
                // 'wrapper' => [
                    // 'href' => url($this->crud->route . '/email'),
                    // 'class' => 'btn btn-sm btn-link',
                    // 'target' => '_blank',
                // ],
            // ]);

        });
    }

    /**
     * Method to handle the GET request and display the View with a Backpack form
     *
     * @param  int  $id
     * @return \Illuminate\Contracts\View\View
     */
    public function getEmailForm(int $id = null)
    {
        $this->crud->hasAccessOrFail('Email');

        return $this->getForm($id);
    }

    /**
     * Method to handle the POST request and perform the operation
     *
     * @param  int  $id
     * @return array|\Illuminate\Http\RedirectResponse
     */
    public function postEmailForm($id = null)
    {
        $this->crud->hasAccessOrFail('Email');

        // You logic goes here...
        dd('got to ' . __METHOD__, request()->all());

        return true;
    }
}
```


## HOW

### How did you achieve that, in technical terms?

- added a new trait `HasForm`, that custom operations can use to quickly do stuff they needed to do anyway (create the GET and POST routes, show the form);
- added a `quick.blade.php` button that is PERFECT for both form use cases (general-purpose form with top button and item form with line button);
- added a new `inc/form.blade.php` view, which is a general-purpose version of `create.blade.php` and `update.blade.php` (those had quite a few things hardcoded);


### Is it a breaking change?

Completely non-breaking. But it _can_ be turned into a breaking change, by:
- 🟡 changing the Create and Update operations to use this new `form.blade.php` (nice!);
- ⚫️ changing the Create and Update operations to use he `HasForm` trait (probably overkill);
- 🔵 moving `form_content.blade.php` to `inc/form_content.blade.php` so that the `crud` directory only contains direct operation views; all partials would then live in `inc`;

### How can we test the before & after?

Before - the steps in https://backpackforlaravel.com/articles/tutorials/how-to-create-a-custom-operation-that-uses-backpack-fields

After - the steps outlined above

tabacitu commented:
TODOs (keeps it non-breaking):
- 🔴 what do we do about save actions? maybe custom forms should just have "save and back" by default;
- 🔴 are any config values set up by default? should they?
- 🔴 if the developer creates a config file, what name should it have? does the operation really pick it up?
- 🔴 I noticed the operation names have the first letter in capital by default... why? is that standard?
- 🔴 test multiple likely scenarios to see that it fits them well;
- 🟡 can we make `setupEmailRoutes()` shorter or easier to write?
- 🟡 can we make `setupEmailDefaults()` shorter or easier to write?
- 🔵 create a new command to generate a form operation - `php artisan backpack:crud-form-operation Email`;

TODOs (turns it into breaking):
- 🟡 change the Create and Update operations to use this `new form.blade.php`;
- ⚫️ change the Create and Update operations to use the `HasForm` trait (probably overkill);
- 🔵 move `form_content.blade.php` to `inc/form_content.blade.php` so that the `crud` directory only contains direct operation views; all partials would then live in `inc`;

tabacitu commented on src/app/Http/Controllers/Operations/Concerns/HasForm.php:
```diff
@@ -0,0 +1,52 @@
+     * @param  int  $id
+     * @return \Illuminate\Contracts\View\View
+     */
+    public function getForm($id = null)
```
Once HasForm is used, this will be a method on the CrudController... so maybe the name is too general? Might conflict with custom methods.

tabacitu commented on src/app/Http/Controllers/Operations/Concerns/HasForm.php:
```diff
@@ -0,0 +1,52 @@
+
+trait HasForm
+{
+    protected function formRoutes(string $operationName, bool $routesHaveIdSegment, string $segment, string $routeName, string $controller): void
```
Once HasForm is used, this will be a method on the CrudController... so maybe the name is too general? Might conflict with custom methods.

tabacitu commented on src/app/Library/CrudPanel/Traits/Fields.php:
```diff
@@ -344,8 +344,6 @@ public function getFields()
      * Check if the create/update form has upload fields.
      * Upload fields are the ones that have "upload" => true defined on them.
      *
-     * @param  string  $form  create/update/both - defaults to 'both'
-     * @param  bool|int  $id  id of the entity - defaults to false
```
Note: these arguments have disappeared a long time ago, we just never removed the docblocks.

tabacitu commented:
@pxpm I want to finish this up tomorrow morning. Do you see any reason this approach wouldn't work well for some realistic 80% use cases? What do you think overall? Any concerns?

pxpm commented:
Hey Cristian.

This is very similar to the approach I started but I was going a little bit further, and I think here we will need to go a little bit further too to fix some issues or provide a little better DX.

Let's talk about a real scenario:

`BanOperation` - An operation that allow you to ban entries, in this case **users**. 
- It should provide a "quick ban" inline button that will ask only for the reason and ban the user for 1 week. 
- a general form to "ban users" where admin can select one or more users and provide the reason and duration of the ban. (top button) 
- an inline button to view the ban history of a specific user (no form)

So I am assuming here that the BanOperation will have multiple forms, and not **only one form**, in that regard I think the routing/form definition could be improved, just not 100% sure  how:

I was looking to define something like this for routes:
```php
trait BanOperation
{
    use HasForm;

    protected function setupBanRoutes($segment, $routeName, $controller)
    {
        // this make sense for forms, so can be provided by HasForms
        // the first the form to ban a single user:
        $this->setupFormRoute('ban-system', 'id'); // admin/user/ban-system/{id}/ban [GET/POST]
        // and then a general "ban users" form
        $this->setupFormRoute('ban-system');// admin/user/ban-system/ban [GET/POST]

        // for "consistency" we can optionally have a similar approach for regular urls that work in 80%+
        $this->setupRoute('history', 'id'); // admin/user/history/{id}/ban [GET]

        // as examples
        $this->setupRoute('ban-system/ban-history', 'id', 'POST'); // admin/user/ban-system/{id}/ban [POST]
        $this->setupRoute('ban-system'); // admin/user/ban-system/ban [GET]

        // notice the usage of named arguments to omit some of them
        $this->setupRoute('ban-system', method: 'POST');
        
        // any complex scenario can still use Route facade directly
        Route::post();
    }
}
```
We maybe use a little bit of PHP magic here to ease developer life a bit and automatically get the function parameters with `debug_back_trace()` ($segment, $routeName, $controller)  in the setup routes methods without explicitly passing them?

Imagine that this is all good and working, we got the routes setup. AFAIK, this is a non-bc.

Next thing in line is the configurations, mainly the default ones. 
- save actions: I agree with you, we should only probably have 1 save action in the form.
- as a developer I expect to create a `config/backpack/operations/ban.php` and that it would be picked as my config file.  

```php
protected function setupBanDefaults()
    {
        // what users can do in this operation
        CRUD::allowAccess(['view-user-bans', 'ban-users', 'ban-user']);

        // I don't expect to be creating the "save form button" etc, so the HasForms trait can provide some initialization function for some simple defaults:
        $this->setupForms();

        //translates to :
        $this->crud->operation('ban', function () {
            //load my config/backpack/operations/ban.php
            $this->crud->loadDefaultOperationSettingsFromConfig();
            
            // the save action should probably be automatically added by `setupParameterFormRoute/setupFormRoute` functions, 
            // as the save action href is dependant on the form routes.
        });

        // add the operation stuff, buttons, columns etc
        $this->crud->operation('list', function () {
            $this->crud->button('ban-user')->stack('line')->view('crud::buttons.quick')->form('ban-user');
            $this->crud->button('ban-users')->stack('top')->view('crud::buttons.quick')->form('ban-users');
            $this->crud->button('view-user-ban')->stack('line')->view('crud::buttons.quick')->url(route('ban.get.history.ban')); //naming routes is a bit confusing here
        });
    }

// we can check if the CrudController is using `HasForms`, in case it is we can register a macro `form()` on the CrudButton ? This will link the button with the "show form url".
```

My ideas are a bit confusing, I am sure they make more sense in my head than when writing, but let's focus on the essential, a "real usecase", because of what I think is missing here is working with a "real operation". 

What you think in general from the above ? Do your implementation/thinking allow us to do something similar ? 

Cheers

tabacitu commented:
Pedro... I think we have something here... I've fixed the above... and did your Ban operation example. And man... it took me just a few minutes to create all 3 of them 🤯 This is super SUPER powerful, what we've done here. You can take a look at the working code for `BanOperation`, `GeneralBanOperation` and `BanHistoryOperation` - https://gist.github.com/tabacitu/aa7b3ecbe2c43c9bb4fa81433f6ba29f - notice I've split up your functionality into 3 distinct operations. Because... that's what they are! Sure you can keep them in one operation, but you have to override a lot of things. Follow the protocol (one form = one operation) and you will do things amazingly fast.

---

After the changes above, here's how an `EmailOperation` looks:
```php
<?php

namespace App\Http\Controllers\Admin\Operations;

use Illuminate\Support\Str;
use Backpack\CRUD\app\Library\CrudPanel\CrudPanelFacade as CRUD;
use Backpack\CRUD\app\Http\Controllers\Operations\Concerns\HasForm;

trait EmailOperation
{
    use HasForm;

    /**
     * Define which routes are needed for this operation.
     *
     * @param string $segment    Name of the current entity (singular). Used as first URL segment.
     * @param string $routeName  Prefix of the route name.
     * @param string $controller Name of the current CrudController.
     */
    protected function setupEmailRoutes($segment, $routeName, $controller)
    {
        $this->formRoutes(
            operationName: 'email',
            routesHaveIdSegment: true,
            segment: $segment,
            routeName: $routeName,
            controller: $controller
        );
    }

    /**
     * Add the default settings, buttons, etc that this operation needs.
     */
    protected function setupEmailDefaults()
    {
        $this->formDefaults(
            operationName: 'email',
            buttonStack: 'line',
            buttonMeta: [
                'icon' => 'la la-envelope',
                'label' => 'Send Email',
            ]
        );
    }

    /**
     * Method to handle the GET request and display the View with a Backpack form.
     *
     * @param  int  $id
     * @return \Illuminate\Contracts\View\View
     */
    public function getEmailForm(int $id = null)
    {
        $this->crud->hasAccessOrFail('email');

        return $this->formView($id);
    }

    /**
     * Method to handle the POST request and perform the action.
     *
     * @param  int  $id
     * @return array|\Illuminate\Http\RedirectResponse
     */
    public function postEmailForm($id = null)
    {
        $this->crud->hasAccessOrFail('email');

        return $this->formAction($id, function($inputs, $entry) {
            // You logic goes here...
            // dd('got to ' . __METHOD__, $inputs, $entry);

            // show a success message
            \Alert::success('Email was sent!')->flash();
        });
    }
}
```

this also needs in the CrudController to have:
```php
    use \App\Http\Controllers\Admin\Operations\EmailOperation;

    protected function setupEmailOperation()
    {
        CRUD::field('to')->type('email')->size(4)->validationRules('required|email');
        CRUD::field('cc')->type('text')->size(4);
        CRUD::field('bcc')->type('text')->size(4);
        CRUD::field('subject')->type('text')->validationRules('required');
        CRUD::field('message')->type('textarea')->validationRules('required');
    }
```

but that's no different from what was before.

---

In total, I've checked:
- ✅ validation works as expected;
- ✅ one save action by default (save_and_back); gets performed automatically by form action;
- ✅ operation names are now `camelCase` as expected;
- ✅ if you create a config file in `config/operations` it will be picked up; if not, the default `backpack.operations.form` will get used (as you did in your branch https://github.com/Laravel-Backpack/CRUD/compare/v6...wip-form-operations )
- ✅ `HasForm` now helps with all fours steps of creating a form operation:
    - ✅ `formRoutes()` helps set up the GET and POST routes;
    - ✅ `formDefaults()` helps set up the config, save action and button;
    - ✅ `formView()` returns the form view (used in GET method);
    - ✅ `formAction()` performs the form validation, model events, then runs the callback the dev gave it to perform whatever custom action is needed of that form, then redirects to the right page using save_action;

---

I'm pretty happy with the state of this. I think it's the 20% that covers 80% of use cases. Any more than this... and we'd be falling down the rabbit hole of developing too much, making it too configurable or catering to use cases are too rare.

Take another look please... try it out if you want... and let me know if anything else triggers your spidey sense.

Thanks!

tabacitu commented:
Update: I've just created a command to generate a form operation too. Now... it's even faster 🤯 It's crazy-fast, actually! Pedro if you're actually testing, please use this too, you'll love it https://github.com/Laravel-Backpack/Generators/pull/191

tabacitu commented:
I am so excited about this... that I've created a short video to show it off to you. THIS. IS. FREAKING. AWESOME!!!

[![Tabacitu showing off the new form operations](https://cdn.loom.com/sessions/thumbnails/3bf2da4bc4a549cea38518ba4184ebc9-with-play.gif)](https://www.loom.com/share/3bf2da4bc4a549cea38518ba4184ebc9)

pxpm commented:
@tabacitu supa-dupa work 🙏 

I am just "transfering" here what I said to you on discord, and where I think our ideas diverge here:
> notice I've split up your functionality into 3 distinct operations. Because... that's what they are! Sure you can keep them in one operation, but you have to override a lot of things. Follow the protocol (one form = one operation) and you will do things amazingly fast

I know at the moment "this is what is happening", and it's valid to have 1 - 1 form/operation. But I also think that rationale is the culprit of some issues we have. 

For example, InlineCreate should be part of the Create, with just a setting: `allowInlineCreate => true/false` in the operation. They are "technically" the same operation but with two different interfaces. Having them separated led to issues like defining the InlineCreate before Create breaks havoc. 

Let's get back to the "real example":
The `BanOperation` is now split into three operations like you said. You are using the BanOperation in `users`, `tags`, `books`, maybe something else.
If you now want to add the `shadow ban` functionality enabled by default you need to change the operation and **all cruds that use it** to add the `ShadowBanOperation`. 

This gets worst if you have your operation as a composer package to use in multiple projects, you will need to change all cruds x multiple projects after a composer update. 

Imagine the nightmare two weeks later to add the UnbanOperation that requires a reason too.

Bare with me, it works the way we have it at the moment, I just don't think is "fewer work" as it would be if our operations were ready to deal with those scenarios. 

What will happen most of the times is that you will just create the rest of the forms manually, so you will endup with a `formRoutes()` for "the form" and then the additional `Route::get/post` for other stuff that you think relates to that operation. 

```php
protected function setupBanRoutes($segment, $routeName, $controller)
    {
        $this->formRoutes(
            operationName: 'ban',
            routesHaveIdSegment: true,
            segment: $segment,
            routeName: $routeName,
            controller: $controller
        );
        Route::get($segment.'/{id}/unban', [
            'as'        => $routeName.'.unbanForm',
            'uses'      => $controller.'@unbanForm',
            'operation' => 'ban',
        ]);
        Route::put($segment.'/{id}', [
            'as'        => $routeName.'.unban',
            'uses'      => $controller.'@unban',
            'operation' => 'ban',
        ]);
    }
```

Or you create the `UnbanOperation` totally fair, it just has caveats. 

Cheers

tabacitu commented:
> For example, InlineCreate should be part of the Create, with just a setting: allowInlineCreate => true/false in the operation. They are "technically" the same operation but with two different interfaces. Having them separated led to issues like defining the InlineCreate before Create breaks havoc.

`InlineCreate` is PRO, `Create` is free. No way they can be part of the same operation.

> The BanOperation is now split into three operations like you said. You are using the BanOperation in users, tags, books, maybe something else. If you now want to add the shadow ban functionality enabled by default you need to change the operation and all cruds that use it to add the ShadowBanOperation.

**You can bundle all 3 operations into a mega-operation 😀 `BanOperations` can contain all 3 operations.** When you want to add `ShadowBan`, you add create that operation trait, then add it there. Easy-peasy-japanesy. The only thing you DO NOT have is one "access" for all operations inside it. You are forced into more granularity when it comes to access, which is unnecessary in most cases. But it's not a huge tradeoff if you ask me.

> What will happen most of the times is that you will just create the rest of the forms manually, so you will endup with a formRoutes() for "the form" and then the additional Route::get/post for other stuff that you think relates to that operation.

Sure, you can do that too. And that's fine. You can also copy-paste code from HasForm if you want. And that's fine too. **This isn't meant to solve all your problems. This feature is only meant to solve one problem - creating an operation with a form**.

---

**I understand what you're proposing. It _would be cool_ to be able to have an operation with multiple forms. But it would be just that... cool. It's a COULD-HAVE.** When you can easily create one trait that contains as many operations as you want - that gives you 90% of the benefit with 0% cost.

So let's analyze what this COULD-HAVE implies:
- change to `formRoutes()` signature (to add `$actionName` or something like that);
- change to route (it has one more parameter now);
- change to configs (all configs we have now need to be in an array);

All of those are either a DX downgrade (if we force them on developers who only need one form) or clutter in our code. Plus, this **change in scope** (⚠️) brings all sorts of dilemmas we will have to face, to account for the added complexity. Just think of the mess we'll be having with the configs. My point here is: This added complexity _is not worth it_. If a COULD-HAVE means added complexity, it's most likely not worth it.

----

**This is exactly the kind of thing I'm trying to force us NOT to do any more. We DO NOT increase the scope of a task, for a COULD-HAVE. No mas!** Not trying to be a hypocrite here, I do this sometimes too - you know I do. But it's up to me to tell you when you're doing it. And up to you to tell me when I'm doing it. We're accountability buddies in this. **Let's stop adding unnecessary complexity!**

And another thing I'm trying to make us not do any more: **we DO NOT let issues baloon any more**. I've estimated this task to be an M. To take me one week from start to finish. If we change the scope of this task, from "_create an operation with a form_" to "_create an operation with one form or multiple forms_", this will no longer be an M. It won't exactly be an L either, but it will be 2 Ms, most likely. Which is simply not doable - I have other priorities next week, that are FAR more important than this.

So please tell me. What do you think is smarter for us to do? 
(A) **Merge this feature like it was first defined?**
(B) **Change the scope and increase its duration?**

You see where I'm getting at here Pedro? Let's please stop doing this. Let's please stop saying "_oh but what if this also..._" and change the scope of a task. That is bad. That is preventing us from _shipping features_. It's making us _slow_. Let's stop being _slow_ and be _fast_ instead 😅 

You think it's a good idea to make it easier for developers to create multi-form operations? Open an issue for it please. We can just as well create a `HasMultipleForms` trait that helps them do that. Bring up the idea (that's the perfect moment to do so) but... as a separate task, mindful of the fact that we cannot and should not do it, because 
- it changes the scope
- it baloons the task
- it has a different priority than the original task (COULD-DO)

----

Now let's focus on the task at hand, with its defined scope. We set up to do one thing - _make it easier for devs to create an operation with a form_. To create their own operation like Create and Update. We did it, right? If you have no other objections, or found any bugs, I want to merge this 🎉 

**Can I merge this?**

pxpm commented on src/app/Http/Controllers/Operations/Concerns/HasForm.php:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * Set up a GET and POST route, for an operation that contains a form.
+     *
+     * @param  string  $operationName
+     * @param  bool  $routesHaveIdSegment
+     * @param  string  $segment
+     * @param  string  $routeName
+     * @param  string  $controller
+     * @return void
```
```suggestion
```

pxpm commented on src/app/Http/Controllers/Operations/Concerns/HasForm.php:
```diff
@@ -0,0 +1,144 @@
+     * Set up the default configurations, save actions and buttons
+     * for a standard operation that contains a form.
+     *
+     * @param  string  $operationName
+     * @param  string  $buttonStack  line/top/bottom
+     * @param  array  $buttonMeta
+     * @return void
+     */
+    protected function formDefaults(string $operationName, $buttonStack = 'line', $buttonMeta = [])
```
```suggestion
     */
    protected function formDefaults(string $operationName, string $buttonStack = 'line', array $buttonMeta = []) : void
```

pxpm commented on src/app/Http/Controllers/Operations/Concerns/HasForm.php:
```diff
@@ -0,0 +1,144 @@
+        $this->crud->operation($operationName, function () use ($operationName) {
+            // if the backpack.operations.{operationName} config exists, use that one
+            // otherwise, use the generic backpack.operations.form config
+            if (config('backpack.operations.'.$operationName) != null) {
```
```suggestion
            if (config()->has('backpack.operations.'.$operationName)) {
```

pxpm commented on src/app/Http/Controllers/Operations/Concerns/HasForm.php:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * Method to handle the GET request and display the View with a Backpack form.
+     *
+     * @param  int  $id
+     * @return \Illuminate\Contracts\View\View
+     */
+    public function formView($id = null)
```
```suggestion
     */
    public function formView(?int $id = null) : \Illuminate\Contracts\View\View
```

pxpm commented on src/app/Http/Controllers/Operations/Concerns/HasForm.php:
```diff
@@ -0,0 +1,144 @@
+     * It performs the validation, authorization and redirect according to the save action.
+     * But it does NOT save the data. That should be done in the given $formLogic callback.
+     *
+     * @param  int  $id
+     * @param  callable  $formLogic
+     * @return array|\Illuminate\Http\RedirectResponse
+     */
+    public function formAction($id = null, callable $formLogic)
```
```suggestion
     * @return array|\Illuminate\Http\RedirectResponse
     */
    public function formAction(?int $id = null, callable $formLogic)
```

pxpm commented on src/config/backpack/operations/form.php:
```diff
@@ -0,0 +1,40 @@
+    'autoFocusOnFirstField' => true,
+
+    // Where do you want to redirect the user by default, save?
+    // options: save_and_back, save_and_edit, save_and_new
```
```suggestion
```

pxpm commented on src/config/backpack/operations/form.php:
```diff
@@ -0,0 +1,40 @@
+
+    // When the user chooses "save and back" or "save and new", show a bubble
+    // for the fact that the default save action has been changed?
+    'showSaveActionChange' => true, //options: true, false
```
Since there is only one save action, shouldn't we make this false ?

tabacitu replied:
Sure!

```suggestion
    'showSaveActionChange' => false, //options: true, false
```

pxpm commented:
We should probably do some validation between stack / routesHaveIdSegment ?

I just ran the command to generate a new operation with a form, and added a TOP button. 
Since it generates with `routesHaveIdSegment => true`, the top button don't work (method not allowed exception) as there is no `id` if not used in the rows (line button). 

To fix it, I need to either change `routesHaveIdSegment => false` or add the button to the line. 
Since there seems to be a "connection" between the property and the button stack, shouldn't we validate it ? The error is not quite clear of the issue and may be confusing. https://flareapp.io/share/NPGEW395

Apart from that it seems to be working, I've also reviewed Generators PR, there is a bug at the moment, the generated stub does not allow to create forms with `routesHaveIdSegment = false` https://github.com/Laravel-Backpack/Generators/pull/191#discussion_r1232568857

 Good job Cristian 🙏
